### PR TITLE
Document cache.h interfaces based on clockcache comments.

### PR DESCRIPTION
**Pull request by jonhnet**
I tried to document all the public methods on the cache.h polymorphic interface. Please see XXXs for places where I don't even have a guess. But also please check that the other text I wrote is actually correct.

### CODE REVIEW

File: src/cache.h [hunk 1]
```diff
@@ -270,211 +330,541 @@ cache_ctxt_init(cache            *cc,
...
    return cc->ops->page_unpin(cc, page);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * cache_page_sync
+ *
+ * Asynchronously writes the page back to disk. Currently there is no way to
+ * check when the writeback has completed,
+ * XXX thus this call is only useful to prepare a page for eviction?
```
Comment (ajhconway) [inline src/cache.h]: This is used to sync log pages opportunistically. The current API doesn't inform the user when this happens. It's not the ideal API.

File: src/cache.h [hunk 2]
```diff
@@ -270,211 +330,541 @@ cache_ctxt_init(cache            *cc,
...
+/*
+ *-----------------------------------------------------------------------------
+ * cache_extent_sync
+ *
+ * Asynchronously syncs the extent beginning at addr.
+ *
+ * *pages_outstanding is immediately incremented by the number of pages
+ * issued for writeback (the non-clean pages of the extent); as writebacks
+ * complete, *pages_outstanding is decremented. The caller may use the
+ * XXX atomically-read? Because it's an aligned uint64?
```
Comment (ajhconway) [inline src/cache.h]: I think it's assumed that it's an aligned uint64 so on x86 it should be safe to read.

File: src/cache.h [hunk 3]
```diff
@@ -270,211 +330,541 @@ cache_ctxt_init(cache            *cc,
...
+ * value of *pages_outstanding to determine when the extent writeback is
+ * complete.
+ *
+ * All pages in the extent must be clean or cleanable.
+ * XXX What other states might there be, and what would happen if you
```
Comment (ajhconway) [inline src/cache.h]: Pages cannot be write locked for example.

If they were, I think those pages would not be sync'd and I'm not 100% sure what happens.

File: src/cache.h [hunk 4]
```diff
@@ -270,211 +330,541 @@ cache_ctxt_init(cache            *cc,
...
+ * called this with a page in such a state?
+ * XXX should we copy this remark from flush?
+ * XXX Asserts that there are no pins, read locks, claims or write locks.
```
Comment (ajhconway) [inline src/cache.h]: No, I think it's fine if there are pins, read locks or claims, since those do not block writeback.

File: src/cache.h [hunk 5]
```diff
@@ -270,211 +330,541 @@ cache_ctxt_init(cache            *cc,
...
 cache_evict(cache *cc, bool ignore_pinned_pages)
 {
    return cc->ops->evict(cc, ignore_pinned_pages);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * cache_cleanup
+ *
+ * XXX ?
```
Comment (ajhconway) [inline src/cache.h]: Bad naming.

This function performs cache callbacks. Generally this isn't required to be called externally, but there are some cases where it must be. For example, prior to shutdown, calling this ensures that all callbacks are called.

File: src/cache.h [hunk 6]
```diff
@@ -270,211 +330,541 @@ cache_ctxt_init(cache            *cc,
...
 {
    return cc->ops->assert_ungot(cc, addr);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * cache_assert_free
+ *
+ * Debugging facility.
+ * Asserts that no thread has a write lock on the page at addr.
```
Comment (ajhconway) [inline src/cache.h:653]: Needs a rename.

File: src/cache.h [hunk 7]
```diff
@@ -270,211 +330,541 @@ cache_ctxt_init(cache            *cc,
...
+ */
 static inline void
 cache_reset_stats(cache *cc)
 {
    return cc->ops->reset_stats(cc);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * cache_reset_stats
```
Comment (ajhconway) [inline src/cache.h:706]: cache_io_stats

File: src/cache.h [hunk 8]
```diff
@@ -270,211 +330,541 @@ cache_ctxt_init(cache            *cc,
...
 cache_present(cache *cc, page_handle *page)
 {
    return cc->ops->cache_present(cc, page);
 }
 
+/*
+ *-----------------------------------------------------------------------------
+ * cache_enable_sync_get
+ *
+ * XXX No idea.
```
Comment (ajhconway) [inline src/cache.h]: This is a safety function. There is a known deadlock with async gets, so they are (were?) disabled by default. This function must be called to enable their use or perhaps an assertion is triggered.

**Review by ajhconway — COMMENTED** [review]
Thank you for doing this!

I tried to fill in the XXXs. I didn't know 100% all the answers.

I corrected a couple other things too.

File: src/cache.h (hunk 8 above)
Comment (jonhnet) [reply src/cache.h]: Wait, this function enables *sync* gets, not async gets?

Comment (jonhnet) [reply src/cache.h]: Is the known deadlock between a sync get and an async get, or between any two async gets? Or something else?

**Comment by jonhnet** [comment]
@aconway: thanks for all the clarifications. All of these comments make and pass the tests. :v) Please take another peek.

**Comment by jonhnet** [comment]
@ajhconway gentle ping on this collection of comments. Thanks!

**Review by ajhconway — APPROVED** [review]
Ship it

**Pull request merged**